camel/agents/chat_agent.py

```diff
@@ -271,6 +280,101 @@ def step(
+                                       role_type=self.role_type,
+                                       meta_dict=dict(),
+                                       content=choice["message"]['content'])
+            output_messages.append(chat_message)
+        return output_messages
+
+    def handle_stream_response(
+        self,
+        response: Any,
+        num_prompt_tokens: int,
+    ) -> Tuple[List[BaseMessage], List[str], Dict[str, int], str]:
+        r"""
+
+        Args:
+            response (dict): Model response.
+            num_prompt_tokens (int): Number of input prompt tokens.
+
+        Returns:
+            tuple: A tuple of list of output `ChatMessage`, list of
+                finish reasons, usage dictionary, and response id.
+        """
+        content_dict = defaultdict(lambda: "")
+        finish_reasons = defaultdict(lambda: "")
+        output_messages: List[BaseMessage] = []
+        response_id: str = ""
+        # All choices in one response share one role
+        role: str = ""
+        for i, chunk in enumerate(response):
+            chunk: Dict[str, Any]
+            response_id = chunk["id"]
+            for _, choice in enumerate(chunk["choices"]):
+                index = choice["index"]
+                choice: Dict[str, Any]
+                delta = choice["delta"]
+                if len(delta) != 0:
+                    # When response has not been stopped
+                    # Notice that only the first chunk has the "role"
+                    role = delta.get("role", role)
+                    delta_content = delta.get("content", "")
+                    content_dict[index] += delta_content
+                else:
+                    finish_reasons[index] = choice["finish_reason"]
+                    chat_message = BaseMessage(role_name=self.role_name,
+                                               role_type=self.role_type,
+                                               meta_dict=dict(),
+                                               content=content_dict[index])
+                    output_messages.append(chat_message)
+        finish_reasons = [
+            finish_reasons[i] for i in range(len(finish_reasons))
+        ]
+        usage_dict = self.get_usage_dict(output_messages, num_prompt_tokens)
+        return output_messages, finish_reasons, usage_dict, response_id
+
+    def get_usage_dict(self, output_messages: List[BaseMessage],
+                       num_prompt_tokens: int) -> Dict[str, int]:
+        r"""Get usage dictionary when using the stream mode.
+
+        Args:
+            output_messages (list): List of output messages.
+            num_prompt_tokens (int): Number of input prompt tokens.
+
+        Returns:
+            dict: Usage dictionary.
+        """
+        encoding = get_model_encoding(self.model.value_for_tiktoken)
+        completion_tokens = 0
+        for message in output_messages:
+            completion_tokens += len(encoding.encode(message.content))
```

Comment:
What is the difference between this and `num_tokens_from_messages`?

Reply:
It seems that openai usage for completion tokens does not include `tokens_per_message` etc in the `num_tokens_from_messages`